### PR TITLE
Fix gs_frame spawning on monitor other than the one the main window is on

### LINKED ISSUE #4019: Render window is created on secondary monitor
This only happens if I happen to have my cursor/am using an application on my secondary screen. I'd prefer it if the window was created on the same monitor rpcs3 is on, or at least my primary screen.

### PR DESCRIPTION
**Pull request by TGEnigma**
It only took me an hour to figure out that setting the QScreen does nothing 'by design'.
Fixes #4019

### CODE REVIEW

**Comment by Megamouse** [comment]
this makes the gs_frame spawn cut off when rpcs3 is anywhere near the screen borders

File: rpcs3/rpcs3_app.cpp [hunk 1]
```diff
@@ -168,30 +168,38 @@ void rpcs3_app::InitializeCallbacks()
 
 		bool disableMouse = guiSettings->GetValue(gui::gs_disableMouse).toBool();
 
+		// Make display geometry rectangle for the gs frame
+		// Take area of main window & center it neatly inside of it
+		// This prevents the frame being spawned wherever your cursor is, which is really annoying for multi-screen users
+		QRect main_window_geometry = RPCS3MainWin->geometry();
+		QRect gs_frame_geometry = QRect(main_window_geometry.left() + ((main_window_geometry.width() - w) / 2),
+										main_window_geometry.top() + ((main_window_geometry.height() - h) / 2),
```
Comment (Megamouse) [inline rpcs3/rpcs3_app.cpp]: btw, using many tabs misaligns on any other setup
maybe do it like this:
image

File: rpcs3/rpcs3qt/gs_frame.h [hunk 2]
```diff
@@ -20,7 +20,7 @@ class gs_frame : public QWindow, public GSFrameBase
 	bool m_minimized = false;
 
 public:
-	gs_frame(const QString& title, int w, int h, QIcon appIcon, bool disableMouse);
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/gs_frame.h]: we don't need the w and h parameters anymore

File: rpcs3/rpcs3qt/gs_frame.cpp [hunk 3]
```diff
@@ -51,7 +51,7 @@ gs_frame::gs_frame(const QString& title, int w, int h, QIcon appIcon, bool disab
 	m_show_fps = static_cast<bool>(g_cfg.misc.show_fps_in_title);
 
 	resize(w, h);
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/gs_frame.cpp]: the resize can be omited

**Comment by JXPZ** [comment]
Qt: implement simple find dialog It's crashing too hard on this rpcs3 bluid, especially in games.

**Comment by JXPZ** [comment]
https://ci.appveyor.com/api/buildjobs/4ph3usqk88xjj967/artifacts/rpcs3-v0.0.4-2018-01-05-4df69939_win64.zip

**Comment by JXPZ** [comment]
esboco 3
[RPCS3.zip](https://github.com/RPCS3/rpcs3/files/1606350/RPCS3.zip)

**Comment by TGEnigma** [comment]
@Megamouse Should be fixed now.
@JXPZ You should probably make a seperate issue for that.

**Comment by AniLeo** [comment]
@JXPZ Removing spam, stop spamming PRs with unrelated random problems

File: rpcs3/rpcs3_app.cpp [hunk 4]
```diff
@@ -168,36 +170,64 @@ void rpcs3_app::InitializeCallbacks()
...
+		// Get minimum virtual screen x & y for clamping the
+		// window x & y later while taking the width and height
+		// into account, so they don't go offscreen
+		int min_screen_x = std::numeric_limits<int>::max();
+		int max_screen_x = std::numeric_limits<int>::min();
+		int min_screen_y = std::numeric_limits<int>::max();
+		int max_screen_y = std::numeric_limits<int>::min();
+		foreach(auto screen, QApplication::screens())
+		{
+			min_screen_x = std::min(min_screen_x, screen->geometry().x());
```
Comment (Megamouse) [inline rpcs3/rpcs3_app.cpp]: maybe use availableGeometry to exclude the taskbar as well

Comment (TGEnigma) [reply rpcs3/rpcs3_app.cpp]: good call, I'll add it in a bit

**Comment by Megamouse** [comment]
looks good to me. sadly I don't have a second screen right now.

**Comment by TGEnigma** [comment]
I initially added a spacer value that approximates the pixel height of the title bar but I tested it wrong and it was way too small. Big Monitor Problems.
Tweaked size looks perfect for me, I hope it does for you too?

**Comment by Megamouse** [comment]
I asked in the forums. maybe someone knows the answer

**Comment by Megamouse** [comment]
looks perfect now btw

**Comment by Megamouse** [comment]
actually you could use setFramePosition instead of setGeometry. but then you'd have to reconsider the lower and right screenborders again.
I think that would be more crossplat friendly

**Comment by TGEnigma** [comment]
But then I can't center it properly because the size of the border isn't known until it's shown.

**Comment by Megamouse** [comment]
It would be best if they provide that functionality upstream

**Comment by TGEnigma** [comment]
In that case, what do we do for now?

**Comment by Megamouse** [comment]
Just leave it. It's better than before anyway

**Comment by Megamouse** [comment]
the guy in the forum told me to check frameGeometry()
I'm not gonna look into it right now. maybe sometime tomorrow

**Comment by TGEnigma** [comment]
frameGeometry() only works once the window already exists as Qt can't know the actual size before the native window is created

**Comment by Megamouse** [comment]
Ah so it's what we tried already. Sad

File: rpcs3/rpcs3_app.cpp [hunk 5]
```diff
@@ -168,36 +170,65 @@ void rpcs3_app::InitializeCallbacks()
...
+		QRect main_window_geometry = RPCS3MainWin->geometry();
+
+		// Get minimum virtual screen x & y for clamping the
+		// window x & y later while taking the width and height
+		// into account, so they don't go offscreen
+		int min_screen_x = std::numeric_limits<int>::max();
+		int max_screen_x = std::numeric_limits<int>::min();
+		int min_screen_y = std::numeric_limits<int>::max();
+		int max_screen_y = std::numeric_limits<int>::min();
+		foreach(auto screen, QApplication::screens())
```
Comment (danilaml) [inline rpcs3/rpcs3_app.cpp]: use range for

**Comment by danilaml** [comment]
Does something like http://doc.qt.io/qt-5/qwindow.html#setScreen not work?

File: rpcs3/rpcs3_app.cpp (hunk 5 above)
Comment (TGEnigma) [reply rpcs3/rpcs3_app.cpp]: gotcha

**Comment by TGEnigma** [comment]
No it does not. Refer to here:
https://bugreports.qt.io/browse/QTBUG-50391?focusedCommentId=302835&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-302835

**Comment by danilaml** [comment]
@TGEnigma  OK. Any reason why proposed workaround is not sufficient?

**Comment by TGEnigma** [comment]
* Centers on the screen, not on the main window
* Doesn't account for extents of the screen & taskbar
It's just a more complete solution.

File: rpcs3/rpcs3_app.cpp [hunk 6]
```diff
@@ -167,37 +169,40 @@ void rpcs3_app::InitializeCallbacks()
...
+			frame = new gs_frame("Vulkan", frame_geometry, RPCS3MainWin->GetAppIcon(), disableMouse);
+			break;
 		}
 #ifdef _MSC_VER
 		case video_renderer::dx12:
 		{
-			gs_frame* ret = new gs_frame("DirectX 12", w, h, RPCS3MainWin->GetAppIcon(), disableMouse);
-			gameWindow = ret;
-			return std::unique_ptr<gs_frame>(ret);
+			gs_frame* ret = new gs_frame("DirectX 12", frame_geometry, RPCS3MainWin->GetAppIcon(), disableMouse);
```
Comment (scribam) [inline rpcs3/rpcs3_app.cpp]: Shouldn’t be something like `frame = new gs_frame("DirectX 12", ...);`?
Or are you still tring to kill DirectX 12? :smile:

Comment (TGEnigma) [reply rpcs3/rpcs3_app.cpp]: maybe unintentionally :P

File: rpcs3/rpcs3.vcxproj.filters [hunk 7]
```diff
@@ -599,6 +604,9 @@
     <ClCompile Include="QTGeneratedFiles\Debug - LLVM\moc_register_editor_dialog.cpp">
       <Filter>Generated Files\Debug - LLVM</Filter>
     </ClCompile>
+    <ClCompile Include="rpcs3qt\utilities.cpp">
```
Comment (scribam) [inline rpcs3/rpcs3.vcxproj.filters]: I don't see the added file "utilities.cpp" in you commit, is it normal?

**Comment by KDZ12** [comment]
esboco 2
[RPCS3.zip](https://github.com/RPCS3/rpcs3/files/1629225/RPCS3.zip)

**Comment by scribam** [comment]
@KDZ12 Please use the latest version of this PR: https://ci.appveyor.com/api/buildjobs/tahjuf8wc0npoos9/artifacts/rpcs3-v0.0.4-2018-01-14-f29cfa3d_win64.zip
This problem should be solved.

**Comment by KDZ12** [comment]
Thanks friend I am using this version aki rpcs3 https://ci.appveyor.com/api/buildjobs/tahjuf8wc0npoos9/artifacts/rpcs3-v0.0.4-2018-01-14-f29cfa3d_win64.zip

**Comment by KDZ12** [comment]
Bayonetta 

[RPCS3.zip](https://github.com/RPCS3/rpcs3/files/1629267/RPCS3.zip)

**Comment by TGEnigma** [comment]
This PR doesnt touch DX12 so maybe you should be using a backend that isnt broken.

File: rpcs3/rpcs3.vcxproj.filters (hunk 7 above)
Comment (TGEnigma) [reply rpcs3/rpcs3.vcxproj.filters]: I don't know how that could still be there, but I'll remove it.

**Pull request merged**